src/Core/Streaming/Watermark.h

```diff
@@ -9,5 +9,48 @@ namespace Streaming
 /// The 1970-01-01 00:00:00 is Int64 value 0, earlier times are negative, so we need to use signed Int64 as watermark
 constexpr Int64 INVALID_WATERMARK = std::numeric_limits<Int64>::min();
 constexpr Int64 TIMEOUT_WATERMARK = std::numeric_limits<Int64>::max();
+
+
+enum class WatermarkStrategy
+{
+    Unknown,
+
+    /// No watermark
+    None,
+
+    /// Watermark is generated based on each event, this is used for `EMIT AFTER WATERMARK WITHOUT DELAY`
+    /// - Not allow time skew
+    Ascending,
+
+    /// Watermark is generated based on last event with a delay interval, this is used for `EMIT AFTER WATERMARK WITH DELAY <delay_interval>`:
+    /// - Allow time skew in a certain range `[<max_event_ts - delay_interval>, <max_event_ts>]`)
+    BoundedOutOfOrderness,
+
+    /// (Built-in): Watermark is generated based on a batch events:
+    /// - Allow time skew in one batch events
+    OutOfOrdernessInBatch,
+
+    /// (Built-in): Watermark is generated based on new window of a batch events:
+    /// - Allow time skew in same window and one batch
+    OutOfOrdernessInWindowAndBatch,
+
+    /// TODO: more strategies ...
+};
+
+enum class WatermarkEmitMode
```

Comment:
Why we need this since the WatermarkStrategy already says this purpose ?

Reply:
This shall be just `EmitMode` instead of WatermarkEmitMode or better `EmitStrategy`

Reply:
`WatermarkStrategy` tells us how to stamp watermark.
`EmitStrategy` tells us how to do finalization when watermark has been observed.

Reply:
So for EmitStrategy, we only have 4 combinations :
1. `emit all states` in a timed window or global window
2. `emit only changed states per group key` in a timed window or global window
3. `emit changelog`, which works for 1 and 2)

Reply:
So the valid emit strategy for now could be

```
1. emit all states
2. emit only keyed changed states
3. emit changelog for keyed changed states
```

Reply:
Observing the timestamp, stamp watermark, assign window, emit states are all hand in hand things, we may need think about them carefully.

Previously we basically did everything (except emit states) in the stamper with session window as an exception.

I like the idea to separate the  `Observing the timestamp, stamp watermark` and `assigning window` and `emit states`.

Reply:
If we separate watermark stamping and window assignment, watermark itself could be pretty simple and having these use cases
1. Ascending - basically proc time processing (EMIT ON PROCTIME) or row number processing  (row number as a timestamp, per row emit in future). For every event in a chunk, stamp a timestamp for each of them. This seems a stateless watermark strategy (probably we still have state incase wall clock rewind / adjusted on that machine ?)
2. BoundedOutOfOrderness - (EMIT ... WITH DELAY), observing the timestamp column for each event, maintain the max timestamp observed so far, if the out of order for an event is within delay, still assign a watermark to it, update the max observed timestamp as we go.
3. OutOfOrdernessInWindowAndBatch - do we need that ? Since for tumble, hopping window, we know the delay automatically ? it seems a special case of 2 ?

Reply:
For session window, do we need a watermark ?

Reply:
Yes, there are more and more actions. Separating these can better maintain them.
in fact, the `EmitStrategy` should only used in AggregatingTransform.

Reply:
I will separate them in another PR.